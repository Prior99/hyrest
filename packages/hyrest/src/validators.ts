import { Scope } from "./scope";

export interface Validation {
    error?: string;
}

export type Validator<T> = (input: T, scope?: Scope) => Validation | Promise<Validation>;

/**
 * Enforces `input` to be one of the values specified in `options`.
 *
 * @param options The options of which `input` needs to be.
 *
 * @return A validator which checks if the input is one of `options` and returns an error if that was
 *         not the case.
 */
export function oneOf<T>(...options: T[]): Validator<T> {
    return (value: T) => {
        if (typeof value === "undefined") { return {}; }
        if (!options.includes(value)) { return { error: `Not one of (${options.join(", ")}).` }; }
        return {};
    };
}

/**
 * Makes sure the given input is not `null` or `undefined`.
 *
 * @param input The input to check.
 *
 * @return The input if it was not `undefined` or `null`, otherwise an error will be returned.
 */
export function required<T>(value: T): Validation {
    if (typeof value === "undefined" || value === null) { return { error: "Missing required field." }; }
    return {};
}

export function length({ min, max }: { max: number, min: number }): Validator<string | any[]>;
export function length(min: number, max: number): Validator<string | any[]>;
/**
 * Validates the input string to be at least of length `min` and not longer than `max`.
 *
 * @param options An object with the keys `min` and `max` for the minimum and maximum length of the string.
 *
 * @return A validator which checks the length of the input string.
 */
export function length(arg1: { max: number, min: number } | number, arg2?: number): Validator<string | any[]> {
    const { min, max } = typeof arg1 === "object" ? arg1 : { min: arg1, max: arg2 };
    return (value: string | any[]) => {
        if (typeof value === "undefined") { return {}; }
        if (value === null) { return { error: "Cannot determine length of null." }; }
        if (typeof max !== "undefined" && value.length > max) {
            return { error: `Exceeds maximum length of ${max}.` };
        }
        if (typeof min !== "undefined" && value.length < min) {
            return { error: `Shorter than minimum length of ${min}.` };
        }
        return {};
    };
}

// Taken from http://emailregex.com/
const emailValidationRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/; // tslint:disable-line
// Taken from https://github.com/mixer/uuid-validate
const uuidValidationRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-4][0-9a-f]{3}-[0-9a-f]{4}-[0-9a-f]{12}$/;

/**
 * Validates that the input is an email.
 *
 * @param value The value to check.
 *
 * @return An error if `value` was not an email.
 */
export function email(value: string): Validation {
    if (typeof value === "undefined") { return {}; }
    if (typeof value !== "string" || !value.match(emailValidationRegex)) { // tslint:disable-line
        return { error: `String is not a valid email.` };
    }
    return {};
}

/**
 * Limits a validator to a given scope. The validator will only be executed if the scope matches.
 *
 * @param onlyScope The scope to limit the validator to.
 * @param validator The validator to execute if the scope matches.
 *
 * @return A validator only executed if the scope matches.
 */
export function only<T>(onlyScope: Scope, validator: Validator<T>): Validator<T> {
    return (value: T, currentScope: Scope) => {
        if (currentScope === onlyScope) {
            return validator(value);
        }
        return {};
    };
}

export function range({ min, max }: { max: number, min: number }): Validator<number>;
export function range(min: number, max: number): Validator<number>;
/**
 * Validates the input to be a number in the specified range.
 *
 * @param options An object with the keys `min` and `max` for the minimum and maximum length of the string.
 *
 * @return A validator which checks the range of the input string.
 */
export function range(arg1: { max: number, min: number } | number, arg2?: number): Validator<number> {
    const { min, max } = typeof arg1 === "object" ? arg1 : { min: arg1, max: arg2 };
    return (value: number) => {
        if (typeof value === "undefined") { return {}; }
        if (value === null) { return { error: "Cannot compare with null." }; }
        if (typeof max !== "undefined" && value > max) {
            return { error: `Exceeds maximum of ${max}.` };
        }
        if (typeof min !== "undefined" && value < min) {
            return { error: `Less than minimum of ${min}.` };
        }
        return {};
    };
}

/**
 * Validates that the input is an uuid.
 *
 * @param value The value to check.
 *
 * @return An error if `value` was not an uuid.
 */
export function uuid(value: string): Validation {
    if (typeof value === "undefined") { return {}; }
    if (typeof value !== "string" || !value.match(uuidValidationRegex)) {
        return { error: `String is not a valid uuid.` };
    }
    return {};
}
