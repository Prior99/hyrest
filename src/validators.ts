export interface Validation {
    error?: string;
}

export type Validator<T> = (input: T) => Validation | Promise<Validation>;

/**
 * Enforces `input` to be one of the values specified in `options`.
 *
 * @param options The options of which `input` needs to be.
 *
 * @return A Converter which checks if the input is one of `options` and returns an error if that was
 *         not the case and the `input` itself otherwise.
 */
export function oneOf<T>(...options: T[]): Validator<T> {
    return (value: T) => {
        if (typeof value === "undefined") { return {}; }
        if (!options.includes(value)) { return { error: `Not one of (${options.join(", ")}).` }; }
        return {};
    };
}

/**
 * Makes sure the given input is not `null` or `undefined`.
 *
 * @param input The input to check.
 *
 * @return The input if it was not `undefined` or `null`, otherwise an error will be returned.
 */
export function required<T>(value: T): Validation {
    if (typeof value === "undefined" || value === null) { return { error: "Missing required field." }; }
    return {};
}
