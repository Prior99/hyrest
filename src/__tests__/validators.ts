import { oneOf, required, email, length } from "../validators";

[
    {
        options: [1, 2, 3, 5],
        values: [
            0,
            27,
            -1,
            "1",
            "test",
            {},
            [],
            null, //tslint:disable-line
            undefined,
        ],
    },
    {
        options: ["a", "b", "c"],
        values: [
            0,
            27,
            -1,
            "1",
            "test",
            {},
            [],
            null, //tslint:disable-line
            undefined,
        ],
    },
    {
        options: [],
        values: ["a"],
    },
].forEach(({ options, values }) => {
    values.forEach(value => {
        test(`oneOf detects "${value}" as invalid`, () => {
            expect(oneOf(...options)(value)).toMatchSnapshot();
        });
    });
    test("oneOf detects each option as valid", () => {
        options.forEach(option => {
            expect(oneOf(...options)(option)).toEqual({});
        });
    });
});

[
    undefined,
    null, //tslint:disable-line
    {},
    [],
    9,
    "",
].forEach(value => {
    test(`required detects "${value}" as expected`, () => {
        expect(required(value)).toMatchSnapshot();
    });
});

[
    undefined,
    null, //tslint:disable-line
    {},
    [],
    9,
    "",
    "test@",
    "test@example.com",
    "test@example",
].forEach(value => {
    test(`email detects "${value}" as expected`, () => {
        expect(email(value)).toMatchSnapshot();
    });
});

[
    {
        range: { min: 5 },
        tests: ["", "1", "test", "longertext", null, undefined], // tslint:disable-line
    },
    {
        range: { min: 5, max: 10 },
        tests: ["", "1", "test", "longertext", "reallylongtextthatexceeds10", 9],
    },
    {
        range: { max: 10 },
        tests: ["", "1", "test", "reallylongtextthatexceeds10", {}, []],
    },
].forEach(({ range, tests }) => {
    tests.forEach(value => {
        test(`length detects "${value}" as expected with ${range}`, () => {
            expect(length(range)(value)).toMatchSnapshot();
        });
    });
});

test("length with the input as args", () => {
    expect(length(1, 5)("")).toEqual({ error: `String is shorter than 1.` });
    expect(length(1, 5)("test")).toEqual({});
    expect(length(1, 5)("toolong")).toEqual({ error: `String exceeds maximum length of 5.` });
});
