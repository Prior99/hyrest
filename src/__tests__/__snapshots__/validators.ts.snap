// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email detects "" as expected 1`] = `
Object {
  "error": "String is not a valid email.",
}
`;

exports[`email detects "" as expected 2`] = `
Object {
  "error": "String is not a valid email.",
}
`;

exports[`email detects "[object Object]" as expected 1`] = `
Object {
  "error": "String is not a valid email.",
}
`;

exports[`email detects "9" as expected 1`] = `
Object {
  "error": "String is not a valid email.",
}
`;

exports[`email detects "null" as expected 1`] = `
Object {
  "error": "String is not a valid email.",
}
`;

exports[`email detects "test@" as expected 1`] = `
Object {
  "error": "String is not a valid email.",
}
`;

exports[`email detects "test@example" as expected 1`] = `
Object {
  "error": "String is not a valid email.",
}
`;

exports[`email detects "test@example.com" as expected 1`] = `Object {}`;

exports[`email detects "undefined" as expected 1`] = `Object {}`;

exports[`length detects "" as expected with [object Object] 1`] = `
Object {
  "error": "String is shorter than 5.",
}
`;

exports[`length detects "" as expected with [object Object] 2`] = `
Object {
  "error": "String is shorter than 5.",
}
`;

exports[`length detects "" as expected with [object Object] 3`] = `Object {}`;

exports[`length detects "" as expected with [object Object] 4`] = `Object {}`;

exports[`length detects "[object Object]" as expected with [object Object] 1`] = `Object {}`;

exports[`length detects "1" as expected with [object Object] 1`] = `
Object {
  "error": "String is shorter than 5.",
}
`;

exports[`length detects "1" as expected with [object Object] 2`] = `
Object {
  "error": "String is shorter than 5.",
}
`;

exports[`length detects "1" as expected with [object Object] 3`] = `Object {}`;

exports[`length detects "9" as expected with [object Object] 1`] = `Object {}`;

exports[`length detects "longertext" as expected with [object Object] 1`] = `Object {}`;

exports[`length detects "longertext" as expected with [object Object] 2`] = `Object {}`;

exports[`length detects "null" as expected with [object Object] 1`] = `
Object {
  "error": "Cannot determine length of null.",
}
`;

exports[`length detects "reallylongtextthatexceeds10" as expected with [object Object] 1`] = `
Object {
  "error": "String exceeds maximum length of 10.",
}
`;

exports[`length detects "reallylongtextthatexceeds10" as expected with [object Object] 2`] = `
Object {
  "error": "String exceeds maximum length of 10.",
}
`;

exports[`length detects "test" as expected with [object Object] 1`] = `
Object {
  "error": "String is shorter than 5.",
}
`;

exports[`length detects "test" as expected with [object Object] 2`] = `
Object {
  "error": "String is shorter than 5.",
}
`;

exports[`length detects "test" as expected with [object Object] 3`] = `Object {}`;

exports[`length detects "undefined" as expected with [object Object] 1`] = `Object {}`;

exports[`oneOf detects "" as invalid 1`] = `
Object {
  "error": "Not one of (1, 2, 3, 5).",
}
`;

exports[`oneOf detects "" as invalid 2`] = `
Object {
  "error": "Not one of (a, b, c).",
}
`;

exports[`oneOf detects "[object Object]" as invalid 1`] = `
Object {
  "error": "Not one of (1, 2, 3, 5).",
}
`;

exports[`oneOf detects "[object Object]" as invalid 2`] = `
Object {
  "error": "Not one of (a, b, c).",
}
`;

exports[`oneOf detects "-1" as invalid 1`] = `
Object {
  "error": "Not one of (1, 2, 3, 5).",
}
`;

exports[`oneOf detects "-1" as invalid 2`] = `
Object {
  "error": "Not one of (a, b, c).",
}
`;

exports[`oneOf detects "0" as invalid 1`] = `
Object {
  "error": "Not one of (1, 2, 3, 5).",
}
`;

exports[`oneOf detects "0" as invalid 2`] = `
Object {
  "error": "Not one of (a, b, c).",
}
`;

exports[`oneOf detects "1" as invalid 1`] = `
Object {
  "error": "Not one of (1, 2, 3, 5).",
}
`;

exports[`oneOf detects "1" as invalid 2`] = `
Object {
  "error": "Not one of (a, b, c).",
}
`;

exports[`oneOf detects "27" as invalid 1`] = `
Object {
  "error": "Not one of (1, 2, 3, 5).",
}
`;

exports[`oneOf detects "27" as invalid 2`] = `
Object {
  "error": "Not one of (a, b, c).",
}
`;

exports[`oneOf detects "a" as invalid 1`] = `
Object {
  "error": "Not one of ().",
}
`;

exports[`oneOf detects "null" as invalid 1`] = `
Object {
  "error": "Not one of (1, 2, 3, 5).",
}
`;

exports[`oneOf detects "null" as invalid 2`] = `
Object {
  "error": "Not one of (a, b, c).",
}
`;

exports[`oneOf detects "test" as invalid 1`] = `
Object {
  "error": "Not one of (1, 2, 3, 5).",
}
`;

exports[`oneOf detects "test" as invalid 2`] = `
Object {
  "error": "Not one of (a, b, c).",
}
`;

exports[`oneOf detects "undefined" as invalid 1`] = `Object {}`;

exports[`oneOf detects "undefined" as invalid 2`] = `Object {}`;

exports[`required detects "" as expected 1`] = `Object {}`;

exports[`required detects "" as expected 2`] = `Object {}`;

exports[`required detects "[object Object]" as expected 1`] = `Object {}`;

exports[`required detects "9" as expected 1`] = `Object {}`;

exports[`required detects "null" as expected 1`] = `
Object {
  "error": "Missing required field.",
}
`;

exports[`required detects "undefined" as expected 1`] = `
Object {
  "error": "Missing required field.",
}
`;
